src/diagnostics/build.rs

```diff
@@ -22,21 +22,59 @@ pub fn collect_build_diagnostics(
         diagnostics
     });
 
-    let root_document = workspace.documents_by_uri.values().find(|document| {
-        document.data.as_latex().map_or(false, |data| {
-            !document.uri.as_str().ends_with(".aux")
-                && data
-                    .extras
-                    .implicit_links
-                    .log
-                    .iter()
-                    .any(|u| u.as_ref() == build_log_uri)
-        })
-    })?;
+    let root_document_uri = Arc::clone(
+        &workspace
+            .documents_by_uri
+            .values()
+            .find(|document| {
+                document.data.as_latex().map_or(false, |data| {
+                    !document.uri.as_str().ends_with(".aux")
+                        && data
+                            .extras
+                            .implicit_links
+                            .log
+                            .iter()
+                            .any(|u| u.as_ref() == build_log_uri)
+                })
+            })?
+            .uri,
+    );
+
+    let build_log_uri = Arc::clone(&build_log_document.uri);
 
-    let base_path = PathBuf::from(root_document.uri.path());
+    let mut base_path = PathBuf::from(root_document_uri.path());
+    base_path.pop();
     for error in &build_log.errors {
-        let position = Position::new(error.line.unwrap_or(0), 0);
+        let full_path = base_path.join(&error.relative_path);
+
+        let doc = if error.line.is_some() && error.hint.is_some() {
+            workspace.load(full_path.clone()).unwrap_or(None)
+        } else {
+            None
+        };
+
+        let position: Position = (if let Some(doc) = doc {
+            // SAFETY: error.line and error.hint are necessarily Some() if we get here
+            let line = error.line.unwrap();
+            let hint: &String = error.hint.as_ref().unwrap();
+            if let Some(hint_line) = doc.text.lines().nth(line as usize) {
+                hint_line
+                    .find(hint)
+                    .map(|col| Position::new(line, (col + hint.len() - 1) as u32))
```

Comment:
If we want to be sure, we need to convert the column offset using `doc.line_index` first. LSP uses `UTF-16` column offsets, while Rust uses `UTF-8` internally.

As an optimization, one could also use the `line_index` to speed up the lookup of the line, but I don't think this is really necessary here.